### PR TITLE
Implement Shopping Cart Functionality

### PR DESCRIPTION
**Pull request by shreya-arya00**
Pull Request Description:

* Introduces the ShoppingCart class to manage user shopping carts.
* Implements methods to add and remove products from the shopping cart.
* Provides a viewCart() method to display the contents of the shopping cart.
* Enhances the OnlineShoppingSystem by enabling users to interact with their shopping carts efficiently.

### CODE REVIEW

File: SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    public Object viewCart() {
+        System.out.println("Items in your shopping cart:");
+        for (Product item : items) {
+            System.out.println(item.getName() + " - $" + item.getPrice());
+        }
+        return null;
+    }
+
+    public Collection<Object> getItems() {
+    return null;}
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java]: why is this returning null ? . can you please change it to return the items in the shopping card.

File: SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    public double getPrice() {
+        return price;
+    }
+}
+
+// ShoppingCart class representing the user's shopping cart
+class ShoppingCart {
+    private List<Product> items;
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java:33]: items is tightly coupled. can you please make this loosely coupled by using dependency injection. 
In future I might want to use some other data structure . I don't want shopping cart class handle the creation of how we store the items.

File: SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    // Add product to shopping cart
+    public void addProduct(Product product) {
+        items.add(product);
+    }
+
+    // Remove product from shopping cart
+    public void removeProduct(Product product) {
+        items.remove(product);
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java]: handle edge cases in case there are no products in the cart. Use CustomExceptions in the code to let the user of the ShoppingCart know what is the issue.

File: SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    // View contents of shopping cart
+    public Object viewCart() {
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java]: Try avoiding returning an Object type. View cart does not require you to return any thing. you can make it void as well.

File: SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        System.out.println("Items in your shopping cart:");
+        for (Product item : items) {
+            System.out.println(item.getName() + " - $" + item.getPrice());
+        }
+        return null;
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java]: why are you returning null ?

File: SoftwareDesignImplementation/src/test/java/ShoppingCartTest.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ShoppingCartTest {
+
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/test/java/ShoppingCartTest.java:9]: Add some test cases where you are trying to remove the products from an empty cart. you can check if an exception is being thrown. For example you can create a new custom Exception Called EmptyCartException.java

File: SoftwareDesignImplementation/src/test/java/ShoppingCartTest.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        cart.addProduct(product2);
+        cart.removeProduct(product1);
+
+        assertEquals(1, cart.getItems().size());
+        assertFalse(cart.getItems().contains(product1));
+        assertTrue(cart.getItems().contains(product2));
+    }
+
+    @Test
+    public void testViewEmptyCart() {
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/test/java/ShoppingCartTest.java]: You dont need this test case. instead test for checking if deleting from an empty cart throws an exception.

File: SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java (hunk 3 above)
Comment (bhuvanmysoresreenivas) [reply SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java]: return the removed product.

File: SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        this.items = new ArrayList<>();
+    }
+
+    // Add product to shopping cart
+    public void addProduct(Product product) {
+        items.add(product);
+    }
+
+    // Remove product from shopping cart
+    public void removeProduct(Product product) {
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java]: return the removed product.

File: SoftwareDesignImplementation/src/test/java/ShoppingCartTest.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        ShoppingCart cart = new ShoppingCart();
+
+        assertEquals("Items in your shopping cart:\nYour shopping cart is empty.\n", cart.viewCart());
+    }
+
+    @Test
+    public void testViewNonEmptyCart() {
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/test/java/ShoppingCartTest.java]: you can remove viewNonEmptyCart test case.

**Review by bhuvanmysoresreenivas — CHANGES_REQUESTED** [review]

**Review by bhuvanmysoresreenivas — APPROVED** [review]

File: SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+// Product class representing individual products
+class Product {
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java:6]: Create a seperate file  Called Product.java and create separate package where all the entities are present

File: SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        return description;
+    }
+
+    public double getPrice() {
+        return price;
+    }
+}
+
+// ShoppingCart class representing the user's shopping cart
+class ShoppingCart {
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java:32]: Create a seperate class called ShoppingCart Service into a seperate file

File: SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+// Custom Exception for an empty shopping cart
+class EmptyCartException extends RuntimeException {
+    public EmptyCartException(String message) {
+        super(message);
+    }
+}
+
+// Main class to run the program
+public class OnlineShoppingSystem {
```
Comment (bhuvanmysoresreenivas) [inline SoftwareDesignImplementation/src/main/java/OnlineShoppingSystem.java:77]: This File should contain only one Class called OnlineShoppingSystem and should contain only one method main
